Use tracing API in TaskManager

pugnascotia opened the pull request:
Split out from #88443. Part of #84369. Use the tracing API that was
added in #87921 in TaskManager. This won't actually do anything until we
provide a tracer with an actual implementation.

pugnascotia committed:
Use tracing API in TaskManager.

Split out from #88443. Part of #84369. Use the tracing API that was
added in #87921 in TaskManager. This won't actually do anything until we
provide a tracer with an actual implemenation.

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

grcevski reviewed (approved):
LGTM! The failures look unrelated at a first glance.

pugnascotia commented:
> LGTM! The failures look unrelated at a first glance.

Plot twist! They are related.

pugnascotia committed:
Fix issue with response header propagation

Change ThreadContext.newTraceContext() so that the context that it
restores is more careful to restore response headers. It does this by
copying the behaviour in newStoredContext().
